File: src/sonic-eventd/lib/src/eventnotify.cpp
#include "eventnotify.h"
#include <algorithm>
#include <stdarg.h>
#include <stdio.h>
#include <syslog.h>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <stdexcept>
#include <vector>
#include <iostream>
#include <ctime>
#include <ratio>
#include <chrono>
#include "schema.h"
#include "select.h"
#include "dbconnector.h"
#include "redisclient.h"
#include "timestamp.h"

using namespace swss;
using namespace std;
using namespace std::chrono;

namespace alertmgr {

string getActionStr(int action);
string getTimestamp();
EventNotify *EventNotify::m_pInstance = NULL;
uint64_t seq_id = 1;

EventNotify::EventNotify(DBConnector* stateDbConnector): 
    m_statePubsubTable(stateDbConnector, EVENT_PUBSUB_TABLE_NAME) {
    SWSS_LOG_ENTER();
}

EventNotify::~EventNotify() {
}

EventNotify *EventNotify::getInstance()
{
   if (!m_pInstance) {
      DBConnector db("EVENT_DB", 0);
      m_pInstance = new EventNotify(&db);
   }

   return m_pInstance;
}

// timeticks are relative to the Unix Epoch 
string getTimestamp() {
    const auto p1 = system_clock::now();
    return to_string(duration_cast<nanoseconds>(p1.time_since_epoch()).count());
}

string getActionStr(int action) {
    switch (action) {
        case CLEAR_ALARM:
            return string("CLEAR");
        case RAISE_ALARM:
            return string("RAISE");
        case ACK_ALARM:
            return string("ACKNOWLEDGE");
        case UNACK_ALARM:
            return string("UNACKNOWLEDGE");
        default:
            return string();
    }
}

void EventNotify::_send_event(const char *ev_name_p, const char *source_p, int action, const char *fmt, ...)
{
    va_list args;
    char buff[1024];
    memset(buff, 0, sizeof(buff));

    SWSS_LOG_ENTER();

    va_start(args, fmt);
    vsnprintf(buff, sizeof(buff), fmt, args);
    va_end (args);

    string tblkey = string(ev_name_p) + to_string(seq_id++);
    vector<FieldValueTuple> fvs;
    fvs.emplace_back("type-id", string(ev_name_p));
    fvs.emplace_back("text", buff);
    fvs.emplace_back("action", getActionStr(action));
    if (source_p && (strlen(source_p) > 0)) {
        fvs.emplace_back("resource", string(source_p));
    }
    fvs.emplace_back("time-created", getTimestamp());

    m_statePubsubTable.set(tblkey, fvs);
}

};

File: src/sonic-eventd/tests/eventnotify/eventnotify.py
from swsscommon import swsscommon
import time
import enum
import time
import os

class EvAction(enum.Enum):
     NOTIFY = 0
     CLEAR_ALARM = 1
     RAISE_ALARM = 2
     ACK_ALARM = 3
     UNACK_ALARM = 4

class EventNotify:
    def __init__(self):
        """Create a new Event Notify class."""
        self.event_db = swsscommon.DBConnector("EVENT_DB", 0, True)
        self.evntPubTbl = swsscommon.Table(self.event_db, "EVENTPUBSUB")
        self.evProfileTbl = swsscommon.Table(self.event_db, "EVPROFILE")
        self.evntStateTbl = swsscommon.Table(self.event_db, "EVENT_STATS")
        self.almStateTbl = swsscommon.Table(self.event_db, "ALARM_STATS")
        self.sequence_id = 0

    def getTicks(self):
        return (time.time_ns())

    def getActionStr(self, action):
        if action ==  EvAction.CLEAR_ALARM:
            return "CLEAR"
        elif action ==  EvAction.RAISE_ALARM:
            return "RAISE"
        elif action ==  EvAction.ACK_ALARM:
            return "ACKNOWLEDGE"
        elif action ==  EvAction.UNACK_ALARM:
            return "UNACKNOWLEDGE"
        else: 
            return ""

    def logevent(self, name, source, action, message):
        self.sequence_id = self.sequence_id + 1
        key = name + str(self.sequence_id)
        actionStr = self.getActionStr(action)
        fvs = swsscommon.FieldValuePairs([("time-created", str(self.getTicks())), ("type-id", name), ("text", message), ("action", actionStr), ("resource", source)])
        self.evntPubTbl.set(key, fvs)
